app/src/main/java/tech/relaycorp/gateway/domain/StoreParcel.kt

```diff
@@ -0,0 +1,80 @@
+package tech.relaycorp.gateway.domain
+
+import tech.relaycorp.gateway.common.Logging.logger
+import tech.relaycorp.gateway.data.repos.ParcelRepository
+import tech.relaycorp.gateway.data.disk.DiskException
+import tech.relaycorp.gateway.data.disk.DiskRepository
+import tech.relaycorp.gateway.data.model.*
+import tech.relaycorp.relaynet.cogrpc.readBytesAndClose
+import tech.relaycorp.relaynet.ramf.RAMFException
+import tech.relaycorp.relaynet.ramf.RAMFMessage
+import java.io.InputStream
+import java.util.*
+import javax.inject.Inject
+
+class StoreParcel
+@Inject constructor(
+    private val parcelRepository: ParcelRepository,
+    private val diskRepository: DiskRepository
+) {
+
+    suspend fun storeParcel(parcelInputStream: InputStream): Result {
+        val parcelBytes = parcelInputStream.readBytesAndClose()
+        val parcel = try {
+            Parcel.deserialize(parcelBytes)
+        } catch (exc: RAMFException) {
+            logger.warning("Malformed Cargo received: ${exc.message}")
+            return Result.Error.Malformed
+        }
+
+        try {
+            parcel.validate()
+        } catch (exc: RAMFException) {
+            logger.warning("Invalid cargo received: ${exc.message}")
+            return Result.Error.Invalid
+        }
+
+        return storeMessage(parcel, parcelBytes)
+    }
+
+    private suspend fun storeMessage(
+        message: RAMFMessage,
```

Comment:
We probably shouldn't try to make this a generic method: it can only work with parcels.

Reply:
This `StoreParcel` is still WIP, I copied the StoreMessage over from the Courier app and renamed things. But this will be rewritten when it actually becomes used.

Reply:
I can remove it actually from this PR.